### DISCUSSION TITLE
External login providers with virtual users/members

### DISCUSSION THREAD

**Discussion by jbreuer** [umbraco/Umbraco-CMS#12741] [Features and ideas]
Umbraco currently supports external login providers for users/members:
https://our.umbraco.com/documentation/reference/security/external-login-providers/

These still need to exist as users/members inside Umbraco. This can be done with the auto-linking feature: 
https://our.umbraco.com/documentation/reference/security/auto-linking/

The problem with the current solution is that users/members need to exist in 2 systems. Umbraco and the external login provider. This makes it harder to use the external login provider as a source of truth for all users/members. If the external login provider has  thousands of users/members you don't want them inside Umbraco as well. 

So this feature request is about virtual users/members. When you authenticate users/members through external providers, Umbraco could create and authenticate a virtual user/member with proper access rights. This user/member exists only as long as the user session lasts.

**Comment by jbreuer** [comment]
Related topic: https://our.umbraco.com/forum/using-umbraco-and-getting-started//109544-external-login-providers-with-virtual-usersmembers

**Comment by jbreuer** [comment]
I managed to get virtual members working. It is an experiment though. I don't know if it is wise to use this approach on a live environment.

You can read the full blog here: https://www.jeroenbreuer.nl/blog/virtual-members-in-umbraco/

**Comment by bergmania** [comment]
Hi @jbreuer.

Thanks for the good work and blog post.

This is definitely something we want, and have talked about for some time already. So far we just didn't have had time to look into it.

I still think it make sense to support having members in two systems sometimes. This could be if you have to add information to the member, and your can't get that from the external provider. That said, there is definitely also use cases where you only need to rely on virtual members.

**Comment by jbreuer** [comment]
Hi @bergmania,

Nice to hear that you want this too. I also agree that having members in two systems could be useful if you want to add information. It would be great to have both autolink or virtual members as options in Umbraco.

**Comment by aguyfromdenmark** [comment]
Hi @bergmania, 
Is there any update on the status of this wanted feature?

**Reply by bergmania** [reply]
Hi.. No we dont have capacity to work on it currently.

**Comment by AndyButland** [comment]
This has been worked on in https://github.com/umbraco/Umbraco-CMS/pull/22162 - we are planning to include this in 17.4, along with documentation of course that I'll link off the PR whilst it's waiting for review.

**Comment by AleksandrRogov** [comment]
hi @AndyButland,
I am a bit late to this discussion, sadly. Somehow missed it being discussed and only saw it in the release notes.
This is a great feature! Thank you for implementing it. 

Do you think if this feature could be _extended_ to introduce "internal lightweight members"? By looking at the implementation, the main issues with this would be in the `MemberManager` (for Password Reset) and `MemberSignInManager` (getting claims). Member workspace would also stay read-only, which is fine. But then this could potentially avoid problems described in this issue #13768 out of the box without a need to introduce an external provider. But I might have missed something, and thus looking for your opinion.

Thanks!
